gdk4/src/event.rs

```diff
@@ -0,0 +1,527 @@
+            let ret = from_glib(ffi::gdk_event_get_axis(
+                self.to_glib_none().0,
+                axis_use.to_glib(),
+                value.as_mut_ptr(),
+            ));
+            let value = value.assume_init();
+            if ret {
+                Some(value)
+            } else {
+                None
+            }
+        }
+    }
+
+    pub fn get_device(&self) -> Option<Device> {
+        unsafe { from_glib_none(ffi::gdk_event_get_device(self.to_glib_none().0)) }
+    }
+
+    pub fn get_display(&self) -> Option<Display> {
+        unsafe { from_glib_none(ffi::gdk_event_get_display(self.to_glib_none().0)) }
+    }
+
+    pub fn get_event_type(&self) -> EventType {
+        unsafe { from_glib(ffi::gdk_event_get_event_type(self.to_glib_none().0)) }
+    }
+
+    pub fn get_history(&self) -> Vec<TimeCoord> {
+        unsafe {
+            let mut out_n_coords = mem::MaybeUninit::uninit();
+            let ret = FromGlibContainer::from_glib_container_num(
+                ffi::gdk_event_get_history(self.to_glib_none().0, out_n_coords.as_mut_ptr()),
+                out_n_coords.assume_init() as usize,
+            );
+            ret
+        }
+    }
+
+    pub fn get_modifier_state(&self) -> ModifierType {
+        unsafe { from_glib(ffi::gdk_event_get_modifier_state(self.to_glib_none().0)) }
+    }
+
+    pub fn get_pointer_emulated(&self) -> bool {
+        unsafe { from_glib(ffi::gdk_event_get_pointer_emulated(self.to_glib_none().0)) }
+    }
+
+    pub fn get_position(&self) -> Option<(f64, f64)> {
+        unsafe {
+            let mut x = mem::MaybeUninit::uninit();
+            let mut y = mem::MaybeUninit::uninit();
+            let ret = from_glib(ffi::gdk_event_get_position(
+                self.to_glib_none().0,
+                x.as_mut_ptr(),
+                y.as_mut_ptr(),
+            ));
+            let x = x.assume_init();
+            let y = y.assume_init();
+            if ret {
+                Some((x, y))
+            } else {
+                None
+            }
+        }
+    }
+
+    pub fn get_seat(&self) -> Option<Seat> {
+        unsafe { from_glib_none(ffi::gdk_event_get_seat(self.to_glib_none().0)) }
+    }
+
+    pub fn get_surface(&self) -> Option<Surface> {
+        unsafe { from_glib_none(ffi::gdk_event_get_surface(self.to_glib_none().0)) }
+    }
+
+    pub fn get_time(&self) -> u32 {
+        unsafe { ffi::gdk_event_get_time(self.to_glib_none().0) }
+    }
+
+    pub fn triggers_context_menu(&self) -> bool {
+        unsafe { from_glib(ffi::gdk_event_triggers_context_menu(self.to_glib_none().0)) }
+    }
+}
+
+impl fmt::Display for Event {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("Event")
+    }
+}
+
+pub trait EventTypeTrait: StaticType + FromGlibPtrFull<*mut ffi::GdkEvent> + 'static {}
+
+macro_rules! define_event {
+    ($rust_type:ident, $ffi_type:path, $rust_type_offset:expr) => {
```

Comment:
Does the `$rust_type_offset` makes any sense here? I couldn't understand what it means to do from looking at the ParamSpec example

Reply:
That was only needed there because for pspecs the `GType`s have no `get_type()` functions but there is just an array. You don't need that here